src/MonitorProfiler/ProbeInstrumentation/ProbeInjector.cpp

```diff
@@ -0,0 +1,213 @@
+    //
+
+    ILInstr* pInsertProbeBeforeThisInstr = rewriter.GetILList()->m_pNext;
+    ILInstr* pNewInstr = nullptr;
+
+    UINT32 numArgs = static_cast<UINT32>(request.boxingTypes.size());
+
+    //
+    // The below IL is equivalent to: ProbeFunction(uniquifier, new object[] { arg1, arg2, ... })
+    // When an argument isn't supported, pass null in its place.
+    //
+
+    /* uniquifier */
+    pNewInstr = rewriter.NewILInstr();
+    pNewInstr->m_opcode = CEE_LDC_I8;
+    pNewInstr->m_Arg64 = request.uniquifier;
+    rewriter.InsertBefore(pInsertProbeBeforeThisInstr, pNewInstr);
+
+    /* Args */
+
+    // Size of array
+    pNewInstr = rewriter.NewILInstr();
+    pNewInstr->m_opcode = CEE_LDC_I4;
+    pNewInstr->m_Arg32 = numArgs;
+    rewriter.InsertBefore(pInsertProbeBeforeThisInstr, pNewInstr);
+
+    // Create the array
+    pNewInstr = rewriter.NewILInstr();
+    pNewInstr->m_opcode = CEE_NEWARR;
+    pNewInstr->m_Arg32 = corLibTypeTokens.systemObjectType;
+    rewriter.InsertBefore(pInsertProbeBeforeThisInstr, pNewInstr);
+
+    for (UINT32 i = 0; i < numArgs; i++)
+    {
+        // New entry on the evaluation stack
+        pNewInstr = rewriter.NewILInstr();
+        pNewInstr->m_opcode = CEE_DUP;
+        rewriter.InsertBefore(pInsertProbeBeforeThisInstr, pNewInstr);
+
+        // Index to set
+        pNewInstr = rewriter.NewILInstr();
+        pNewInstr->m_opcode = CEE_LDC_I4;
+        pNewInstr->m_Arg32 = i;
+        rewriter.InsertBefore(pInsertProbeBeforeThisInstr, pNewInstr);
+
+        // Load arg
+        ULONG32 typeInfo = request.boxingTypes.at(i);
+        if (typeInfo == (SpecialCaseBoxingTypeFlag | static_cast<ULONG32>(SpecialCaseBoxingTypes::TYPE_UNKNOWN)))
+        {
+            pNewInstr = rewriter.NewILInstr();
+            pNewInstr->m_opcode = CEE_LDNULL;
+            rewriter.InsertBefore(pInsertProbeBeforeThisInstr, pNewInstr);
+        }
+        else
+        {
+            pNewInstr = rewriter.NewILInstr();
+            pNewInstr->m_opcode = CEE_LDARG_S;
+            pNewInstr->m_Arg32 = i;
+            rewriter.InsertBefore(pInsertProbeBeforeThisInstr, pNewInstr);
+
+            // Resolve the boxing token.
+            mdToken boxedTypeToken;
+            IfFailRet(GetBoxingToken(typeInfo, corLibTypeTokens, boxedTypeToken));
+            if (boxedTypeToken != mdTokenNil)
+            {
+                pNewInstr = rewriter.NewILInstr();
+                pNewInstr->m_opcode = CEE_BOX;
+                pNewInstr->m_Arg32 = boxedTypeToken;
+                rewriter.InsertBefore(pInsertProbeBeforeThisInstr, pNewInstr);
+            }
+        }
+
+        // Replace the i'th element in our new array with what we just pushed on the stack
+        pNewInstr = rewriter.NewILInstr();
+        pNewInstr->m_opcode = CEE_STELEM_REF;
+        rewriter.InsertBefore(pInsertProbeBeforeThisInstr, pNewInstr);
+    }
+
+    pNewInstr = rewriter.NewILInstr();
```

Comment:
Can you test instrumentation against a method that has a protected exception region starting at il offset 0?